### DISCUSSION TITLE
Configure `initializationOptions` per project

### DISCUSSION THREAD

**Discussion by rbartlensky** [joaotavora/eglot#1135] [Ideas]
Hi!

I am trying to implement a way to have eglot read a file, like `.dir-locals.el`, but for `initializationOptions`.

I have attempted to implement this in `rustic` ([PR](https://github.com/brotzeit/rustic/pull/469/files)), but the author suggested that maybe this can become an `eglot` feature.

My usecase is that I want to configure rust-analyzer on a per-project basis. Since rust-analyzer can only be configured via `initializationOptions`, there is no way to currently add the desired options into something like `.dir-locals.el`. Do you have any thoughts on this?

**Comment by joaotavora** [comment]
Hello,

If rust-analyzer can indeed only read project-specific data via
`initializationOptions`  (and not something more usual such as
workspaceConfiguration) then your best option is to set
eglot-server-programs directory-locally. In each directory where
you need a different rust-analyzer initialization, set
eglot-server-programs for 'rust-mode' in a way that includes a
:initializationOptions (presumably a slightly different one for
each project).

See the https://joaotavora.github.io/eglot/#Setting-Up-LSP-Servers

Hopefully it is clear what the "JSON object" is there.  It's constructed
in a syntax identical to the one used for workspaceConfiguration,
which is described later in the manual.

João

**Reply by rbartlensky** [reply]
Hi,

Thank you for getting back to me so quickly! One thing I'm unsure of is the "directory-locally" part. Could you elaborate on that please?

My hope is that whatever solution we come up with would benefit not just me, but anyone using `eglot`. I would like to find a way to check the `initializationOptions` for an LSP server into git and have other people that use `eglot` be able to seamlessly get the benefits without having to come up with their own logic to load them.  At the moment, you could achieve something similar if you used `.dir-locals.el`  (but only for `workspaceConfigurations`), hence why I am trying to go down this route.

Robert

**Reply by joaotavora** [reply]
"Directory-locally" means "using `dir-locals.el`"! Use that file to set the variable `eglot-server-programs` like i described (`:initializationOptions` is supported in its syntax, as the manual link explains). Check that it works, then check the file into Git.

**Reply by rbartlensky** [reply]
Ah I've never heard about `dir-locals` before, hence my confusion! Are you suggesting something like:
```
((rustic-mode
  . ((eglot-server-programs . (rustic-mode . ("rust-analyzer" :initializationOptions (...)))))))
```

I've had some success with this, however I had to put in my emacs config that `eglot-server-programs` can be safely overriden by `dir-locals.el`. Is that expected? If I don't add that I just get something like "attempted to use minibuffer while using the minibuffer" which I couldn't get past even with recursive minibuffers.

Thanks again, sorry for the late reply!

**Reply by joaotavora** [reply]
> Ah I've never heard about dir-locals before, 

In your very first sentence to this discussion you named this file!

Yes you need the safety switch. If you add that to your configuration it's a one time thing. 

I don't understand the relation to recursive mini buffers.

I would follow @nemethf advice and try to use workspaceComfiguration instead of this technique.

**Reply by rbartlensky** [reply]
> In your very first sentence to this discussion you named this file!

I thought it's specific to `eglot` that's what I meant. I didn't know you could use `dir-locals` outside of `eglot`.

> Yes you need the safety switch. If you add that to your configuration it's a one time thing.

Good to know!

> I don't understand the relation to recursive mini buffers. 

Doesn't matter too much, since marking `eglot-server-programs` as safe to override doesn't lead to any issues. I mostly brought it up in case you've seen the error before and had some fix for it.

> I would follow @nemethf advice and try to use workspaceComfiguration instead of this technique.

Will do, thanks again for your help!

**Comment by nemethf** [comment]
> Since rust-analyzer can only be configured via `initializationOptions`, 

This is generally not true.  Last time I looked, intializaionOptions was only necessary to set the location of the rustfmt executable.  See https://github.com/joaotavora/eglot/discussions/845

**Reply by rbartlensky** [reply]
The [manual](https://rust-analyzer.github.io/manual.html#configuration) seems to suggest that most config values are passed via `initializationOptions`.

**Comment by ia0** [comment]
@rbartlensky you might be interested by the "Experimental setup" section of https://github.com/joaotavora/eglot/discussions/1214. This is my solution for per project initialization options. It might not be exactly what you want because I'm only interested in dynamically changing the options. In particular I don't persist them, which means that restarting emacs would lose any configuration. Maybe I'll add such feature if I need it, but so far I'm fine doing `C-c c` and updating the cargo invocation each time I need it.
